File: packages/core/src/rules/satisfies-references-exist.ts
import type { RequirementWithSource, ValidationError } from "../types.js";

/**
 * GRD-VALID-004: The tool shall validate that every reference id under a
 * requirement's "links" (e.g. satisfies) exists as another requirement in
 * the project. Unknown ids are reported with the source path and missing id.
 */
export function satisfiesReferencesExist(
  requirements: RequirementWithSource[]
): ValidationError[] {
  const errors: ValidationError[] = [];
  const idSet = new Set(requirements.map((r) => r.id));

  for (const r of requirements) {
    const links = r.links ?? [];
    for (const link of links) {
      const targetId = link.satisfies;
      if (targetId && !idSet.has(targetId)) {
        errors.push({
          path: r.sourcePath,
          message: `Requirement ${r.id} references unknown id: ${targetId}`,
        });
      }
    }
  }

  return errors;
}

File: crates/gitreqd-core/src/rules/satisfies_references_exist.rs
//! GRD-VALID-004: Every reference id under a requirement's "links" (e.g. satisfies)
//! must exist as another requirement in the project.

use std::collections::HashSet;

use crate::types::{RequirementWithSource, ValidationError};

pub fn satisfies_references_exist(requirements: &[RequirementWithSource]) -> Vec<ValidationError> {
    let mut errors = Vec::new();
    let id_set: HashSet<&str> = requirements.iter().map(|r| r.id.as_str()).collect();

    for r in requirements {
        for link in r.links.as_deref().unwrap_or(&[]) {
            if let Some(target_id) = &link.satisfies {
                if !target_id.is_empty() && !id_set.contains(target_id.as_str()) {
                    errors.push(ValidationError::new(
                        r.source_path.display().to_string(),
                        format!("Requirement {} references unknown id: {target_id}", r.id),
                    ));
                }
            }
        }
    }

    errors
}
